packages/java-edition/src/mcfunction/checker/index.ts

```diff
@@ -81,43 +88,69 @@ const entity: core.SyncChecker<EntityNode> = (node, ctx) => {
 	}
 }
 
-const itemStack: core.SyncChecker<ItemStackNode> = (node, ctx) => {
+const itemPredicate: core.SyncChecker<ItemPredicateNode> = (node, ctx) => {
 	if (node.nbt) {
 		nbt.checker.index(
 			'minecraft:item',
 			core.ResourceLocationNode.toString(node.id, 'full'),
+			{ isPredicate: true },
 		)(node.nbt, ctx)
 	}
-	if (node.components) {
-		const groupedComponents = new Map<
-			string,
-			core.PairNode<core.ResourceLocationNode, nbt.NbtNode>[]
-		>()
-
-		node.components!.children.forEach(component => {
-			const componentName = core.ResourceLocationNode.toString(
-				component.key!,
-				'full',
-			)
-
-			if (!groupedComponents.has(componentName)) {
-				groupedComponents.set(componentName, [])
+	if (!node.tests) {
+		return
+	}
+	for (const anyOfTest of node.tests.children) {
+		for (const allOfTest of anyOfTest.children) {
+			for (const test of allOfTest.children) {
+				const key = core.ResourceLocationNode.toString(test.key, 'full')
+				if (ComponentTestExactNode.is(test) && test.value) {
+					nbt.checker.index('minecraft:data_component', key)(
+						test.value,
+						ctx,
+					)
+				} else if (ComponentTestSubpredicateNode.is(test) && test.value) {
+					nbt.checker.index('minecraft:item_sub_predicate', key)(
+						test.value,
+						ctx,
+					)
+				}
 			}
+		}
+	}
+}
 
-			groupedComponents.get(componentName)!.push(component)
-		})
-
-		groupedComponents.forEach((components) => {
-			if (components.length > 1) {
-				components.forEach(component => {
-					ctx.err.report(
-						localize('mcfunction.parser.duplicate-components'),
-						component.key!.range,
-						core.ErrorSeverity.Warning,
-					)
-				})
+const itemStack: core.SyncChecker<ItemStackNode> = (node, ctx) => {
+	if (node.nbt) {
+		const id = core.ResourceLocationNode.toString(node.id, 'full')
+		nbt.checker.index('minecraft:item', id)(node.nbt, ctx)
+	}
+	if (!node.components) {
+		return
+	}
+	const groupedComponents = new Map<string, core.AstNode[]>()
+	for (const pair of node.components.children) {
+		if (!pair.key) {
+			continue
+		}
+		const key = core.ResourceLocationNode.toString(pair.key, 'full')
+		if (!groupedComponents.has(key)) {
+			groupedComponents.set(key, [])
+		}
+		groupedComponents.get(key)!.push(pair.key)
+		if (pair.value) {
+			nbt.checker.index('minecraft:data_component', key)(pair.value, ctx)
+		}
+	}
+	for (const [_, group] of groupedComponents) {
+		if (group.length > 1) {
+			for (const node of group) {
+				ctx.err.report(
+					localize('mcfunction.parser.duplicate-components'),
+					node.range,
+					core.ErrorSeverity.Warning,
```

Comment:
warning because Minecraft won't actually error?

Reply:
Mostly because duplicate keys in other places usually show a warning. Errors are for actual parsing issues.